src/main/scala/io/github/tcdl/msb/MsbResponderActor.scala

```diff
@@ -19,47 +20,54 @@ trait MsbResponderActor extends Actor {
   def msbcontext: MsbContext = Msb(context.system).context
 
   /** Create a payload builder with the given body. */
-  def response(body: Any): Payload.Builder = new Payload.Builder().withBody(body)
+  def response(body: Any): RestPayload.Builder[_, _, _, _] = new RestPayload.Builder().withBody(body)
 
   override def preStart(): Unit = {
     super.preStart()
     responderServer.listen()
   }
 
   override def receive = {
-    case IncomingRequest(payload, responder) =>
-      handleRequest(payload, ResponderImpl(responder))
+    case IncomingRequest(payload, ctx) =>
+      handleRequest(Payload(payload), ResponderImpl(ctx.getResponder))
   }
 
-  private lazy val responderServer = objectFactory.createResponderServer(namespace, new MessageTemplate(), requestHandler)
+  private lazy val responderServer = objectFactory.createResponderServer(namespace, new MessageTemplate(), requestHandler, classOf[RestPayload[_, _, _, _]])
   private def objectFactory = msbcontext.getObjectFactory
 
-  private val requestHandler: RequestHandler = {
-    (payload: Payload, responder: JavaResponder) => self ! IncomingRequest(payload, responder)
+  private val requestHandler = {
+    (payload: RestPayload[_, _, _, _], ctx: ResponderContext) => self ! IncomingRequest(payload, ctx)
   }
 }
 
 object MsbResponderActor {
   type MsbRequestHandler = PartialFunction[(Payload, Responder), Unit]
   case class Ack(timeout: FiniteDuration, remaining: Int = 0)
 
-  private case class IncomingRequest(payload: Payload, responder: JavaResponder)
+  private case class IncomingRequest(payload: RestPayload[_, _, _, _], context: ResponderContext)
+
+  case class Payload(payload: RestPayload[_, _, _, _]) {
+    def getBodyAs[T <: Any](implicit tag: ClassTag[T]): T = convert(payload.getBody)
+    def getBodyBuffer: Array[Byte] = payload.getBodyBuffer
+    def getHeadersAs[T <: Any](implicit tag: ClassTag[T]): T = convert(payload.getHeaders)
+    def getParamsAs[T <: Any](implicit tag: ClassTag[T]): T = convert(payload.getParams)
+    def getQueryAs[T <: Any](implicit tag: ClassTag[T]): T = convert(payload.getQuery)
+    def getStatusCode: Int = payload.getStatusCode
+    def getStatusMessage: String = payload.getStatusMessage
+  }
 }
 
 trait Responder {
-  def ! (payload: Payload)
+  def ! (payload: RestPayload[_, _, _, _])
```

Comment:
Could we avoid using this horrifically ugly type and replace it with our own Payload object?

The cleanest way is probably to hide that type entirely and let the MsbRequester and MsbResponder only handle our own case class (and convert it to a RestPayload when talking to msb-java).